source/mqtt_agent.c

```diff
@@ -295,35 +305,65 @@ static bool isSpaceInPendingAckList( const MQTTAgentContext_t * pAgentContext )
 
 /*-----------------------------------------------------------*/
 
-static bool addAwaitingOperation( MQTTAgentContext_t * pAgentContext,
-                                  uint16_t packetId,
-                                  Command_t * pCommand )
+static MQTTStatus_t addAwaitingOperation( MQTTAgentContext_t * pAgentContext,
+                                          uint16_t packetId,
+                                          Command_t * pCommand )
 {
-    size_t i = 0;
-    bool ackAdded = false;
+    size_t i = 0, unusedPos = MQTT_AGENT_MAX_OUTSTANDING_ACKS;
+    MQTTStatus_t status = MQTTNoMemory;
     AckInfo_t * pendingAcks = NULL;
 
     assert( pAgentContext != NULL );
     assert( pCommand != NULL );
     assert( packetId != MQTT_PACKET_ID_INVALID );
     pendingAcks = pAgentContext->pPendingAcks;
 
-    /* Look for an unused space in the array of message IDs that are still waiting to
-     * be acknowledged. */
+    /* Before adding the record for the pending acknowledgement of the packet ID,
+     * make sure that there doesn't already exist an entry for the same packet ID.
+     * Also, as part of iterating through the list of pending acknowledgements,
+     * find an unused space for the packet ID to be added, if it can be. */
     for( i = 0; i < MQTT_AGENT_MAX_OUTSTANDING_ACKS; i++ )
     {
         /* If the packetId is MQTT_PACKET_ID_INVALID then the array space is not in
          * use. */
-        if( pendingAcks[ i ].packetId == MQTT_PACKET_ID_INVALID )
+        if( ( unusedPos == MQTT_AGENT_MAX_OUTSTANDING_ACKS ) &&
+            ( pendingAcks[ i ].packetId == MQTT_PACKET_ID_INVALID ) )
         {
-            pendingAcks[ i ].packetId = packetId;
-            pendingAcks[ i ].pOriginalCommand = pCommand;
-            ackAdded = true;
+            unusedPos = i;
+            status = MQTTSuccess;
+        }
+        else if( pendingAcks[ i ].packetId == packetId )
```

Comment:
Please either add an empty `else` block for MISRA, or change this `else if` to an `if`